packages/flutter/lib/src/foundation/consolidate_response.dart

```diff
@@ -47,6 +50,7 @@ Future<Uint8List> consolidateHttpClientResponseBytes(
   HttpClientResponse response, {
   bool autoUncompress = true,
   BytesReceivedCallback onBytesReceived,
+  SendChunkEventsController sendChunkEventsController,
```

Comment:
This seems like an awkward API to have a callback and a separate object that controls if the callback gets called or not. I wonder if we can come up with something better here?

Reply:
+1.  This is not something that should be exposed in a public API.

Reply:
Can you clarify your thought process on why this should not be exposed?
Proposed api is pretty straightforward: callback is invoked only when controller allows it do so. Potentially this can be wrapped into a single class (combining callback and a controller), but in current form it keeps compatibility with old api and reduces complexity of introducing another class.

Reply:
It's exposing an implementation detail.  The caller is already able to specify `onBytesReceived` or leave it null - it's not at all clear from an API perspective why they should also have to specify this opaque controller object.  Even if it was wrapped into a common class with one argument, the same would apply.

Reply:
The fact that `onBytesReceived` is specified doesn't necessary mean that callback has to be invoked(https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/painting/_network_image_io.dart#L94). It's only when user used `addListener` on listener produced by `load`(https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/painting/_network_image_io.dart#L46) with non-null `onChunk` listener, then callback has to be invoked.

This is what this controller enables `NetworkImage` class that uses `consolidateHttpClientResponseBytes` to do.

Reply:
`consolidateHttpClientResponseBytes` itself is part of the public API, though, and should have a sensible API as well. As tvolkert said, just from looking at the API of this function its very odd that I have to specify this opaque controller object.

Reply:
Also, from an API perspective, `onBytesReceived` being specified *should* mean that it has to be invoked.  Perhaps we can fix the problem you're trying to solve inside the impl of `_loadAsync()`?

Reply:
I don't immediately see how this can be solved in `_loadAsync`, if you have some particular insight please let me know.
Conceptually what I'm trying to express is that `onBytesReceived` provided by `_loadAsync` is _optional_, and have to be invoked only if user of `ImageStreamCompleter`(`ImageStreamCompleter` encapsulates `_loadAsync`) did register a callback _with_ `onChunk` handler.

Reply:
I also don't see an immediate solution. Maybe this requires a redesign to make this more sensible?